src/components/Auth.js

```diff
@@ -0,0 +1,207 @@
+import {
+  Box, Button, CircularProgress, Typography,
+} from '@material-ui/core';
+import fetchIntercept from 'fetch-intercept';
+import * as jwt from 'jsonwebtoken';
+import Keycloak from 'keycloak-js';
+import { PropTypes } from 'prop-types';
+import React, {
+  createContext, useContext, useEffect,
+  useLayoutEffect, useMemo, useRef,
+} from 'react';
+import { useMutation } from 'react-query';
+import { Route } from 'react-router-dom';
+
+import api from '@/services/api';
+
+const dbRoles = {
+  admin: 'admin',
+  regular: 'regular',
+  readonly: 'readonly',
+};
+
+const keycloak = Keycloak({
+  realm: window._env_.KEYCLOAK_REALM,
+  clientId: window._env_.KEYCLOAK_CLIENT_ID,
+  url: window._env_.KEYCLOAK_URL,
+  realm_access: { roles: [window._env_.KEYCLOAK_ROLE] },
+});
+
+const AuthContext = createContext(undefined);
+
+const useAuth = () => {
+  const state = useContext(AuthContext);
+
+  if (!state) {
+    throw new Error('context provider for AuthContext is missing');
+  }
+
+  return state;
+};
+
+const AuthProvider = (props) => {
+  const { children } = props;
+
+  const [logInOrOut, { isLoading: isAuthenticating, data, error }] = useMutation(
```

Comment:
this might be a silly question, but i've never used "useMutation" before, do you use it even when nothing changes on the api end?

Reply:
I use it whenever I am making a request that causes changes, in this case it changes the login state for the user. If I were to use useQuery here, I would have to add a bunch of options to stop it from automatically running/rerunning and I'd have to give it a key